adapters/rack/lib/opentelemetry/adapters/rack/middlewares/tracer_middleware.rb

```diff
@@ -0,0 +1,248 @@
+                                                kind: :server)
+          end
+
+          # Per https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#http-server-semantic-conventions
+          #
+          # One of the following sets is required, in descending preference:
+          # * http.scheme, http.host, http.target
+          # * http.scheme, http.server_name, net.host.port, http.target
+          # * http.scheme, net.host.name, net.host.port, http.target
+          # * http.url
+          def request_span_attributes
+            {
+              'component' => 'http',
+              'http.method' => env['REQUEST_METHOD'],
+              'http.url' => full_http_request_url,
+              'http.host' => env['HOST'],
+              'http.scheme' => env['rack.url_scheme'],
+              'http.base_url' => base_url # TODO: check attribute naming/semantics
+            }.merge(allowed_request_headers)
+          end
+
+          def full_http_request_url
+            # NOTE: dd-trace-rb has implemented 'quantization'? (which lowers url cardinality)
+            #       see Datadog::Quantization::HTTP.url
+
+            rack_request.url
+          end
+
+          # https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/data-http.md#name
+          #
+          def request_span_name
+            # uri_path_value:
+            env['PATH_INFO']
+          end
+
+          def base_url
+            if rack_request.respond_to?(:base_url)
+              rack_request.base_url
+            else
+              # Compatibility for older Rack versions
+              rack_request.url.chomp(rack_request.fullpath)
+            end
+          end
+
+          def rack_request
+            @rack_request ||= ::Rack::Request.new(env)
+          end
+
+          def record_and_reraise_error(error)
+            # TODO: implement span.set_error?
+
+            # catch exceptions that may be raised in the middleware chain
+            # Note: if a middleware catches an Exception without re raising,
+            # the Exception cannot be recorded here.
+            request_span.set_error(error) unless request_span.nil?
+            raise error
+          end
+
+          def set_attributes_after_request(span, status, headers, _response)
+            span.status = OpenTelemetry::Trace::Status.http_to_status(status)
+            span.set_attribute('http.status_code', status)
+
+            # TODO: set_attribute('http.route', ... if it's available
+            span.set_attribute('http.status_text', ::Rack::Utils::HTTP_STATUS_CODES[status])
+
+            allowed_response_headers(headers).each { |k, v| span.set_attribute(k, v) }
+          end
+
+          # @return Hash
+          def allowed_request_headers
+            {}.tap do |result|
+              Array(allowed_request_header_names).each do |header|
+                rack_header = "HTTP_#{header.to_s.upcase.gsub(/[-\s]/, '_')}"
```

Comment:
Since the allowed headers are static, can we find some way to cache the header name and the attribute name so we don't have to create a new strings every time?